src/main/java/com/gregtechceu/gtceu/integration/jade/provider/RecipeOutputProvider.java

```diff
@@ -59,19 +67,23 @@ protected void write(CompoundTag data, RecipeLogic recipeLogic) {
 
                 ListTag itemTags = new ListTag();
                 for (var item : itemContents) {
-                    var stacks = ItemRecipeCapability.CAP.of(item.content).getItems();
-                    if (stacks.length == 0) continue;
-                    if (stacks[0].isEmpty()) continue;
-                    var stack = stacks[0];
-
                     var itemTag = new CompoundTag();
-                    GTUtil.saveItemStack(stack, itemTag);
-                    if (item.chance < item.maxChance) {
-                        int count = stack.getCount();
-                        double countD = (double) count * recipe.parallels *
-                                function.getBoostedChance(item, recipeTier, chanceTier) / item.maxChance;
-                        count = countD < 1 ? 1 : (int) Math.round(countD);
-                        itemTag.putInt("Count", count);
+                    if (item.content instanceof IntProviderIngredient provider) {
+                        // don't bother rolling output
+                        itemTag = (CompoundTag) JsonOps.INSTANCE.convertTo(NbtOps.INSTANCE, provider.toJson());
+                    } else {
+                        var stacks = ItemRecipeCapability.CAP.of(item.content).getItems();
+                        if (stacks.length == 0) continue;
+                        if (stacks[0].isEmpty()) continue;
```

Comment:
there checks can be merged into
```suggestion
                        if (stacks.length == 0 || stacks[0].isEmpty()) continue;
```